File: library/src/main/java/com/instacart/library/truetime/TrueTime.java
package com.instacart.library.truetime;

import android.content.Context;
import android.os.SystemClock;
import java.io.IOException;
import java.util.Date;
import java.util.Locale;

public class TrueTime {

    private static final String TAG = TrueTime.class.getSimpleName();

    private static final TrueTime INSTANCE = new TrueTime();
    private static final DiskCacheClient DISK_CACHE_CLIENT = new DiskCacheClient();
    private static final SntpClient SNTP_CLIENT = new SntpClient();

    private static float _rootDelayMax = 100;
    private static float _rootDispersionMax = 100;
    private static int _serverResponseDelayMax = 750;
    private static int _udpSocketTimeoutInMillis = 30_000;

    private String _ntpHost = "1.us.pool.ntp.org";

    /**
     * @return Date object that returns the current time in the default Timezone
     */
    public static Date now() {
        if (!isInitialized()) {
            throw new IllegalStateException("You need to call init() on TrueTime at least once.");
        }

        long cachedSntpTime = _getCachedSntpTime();
        long cachedDeviceUptime = _getCachedDeviceUptime();
        long deviceUptime = SystemClock.elapsedRealtime();
        long now = cachedSntpTime + (deviceUptime - cachedDeviceUptime);

        return new Date(now);
    }

    public static boolean isInitialized() {
        return SNTP_CLIENT.wasInitialized() || DISK_CACHE_CLIENT.isTrueTimeCachedFromAPreviousBoot();
    }

    public static TrueTime build() {
        return INSTANCE;
    }

    public void initialize() throws IOException {
        initialize(_ntpHost);
    }

    /**
     * Cache TrueTime initialization information in SharedPreferences
     * This can help avoid additional TrueTime initialization on app kills
     */
    public synchronized TrueTime withSharedPreferencesCache(Context context) {
        DISK_CACHE_CLIENT.enableCacheInterface(new SharedPreferenceCacheImpl(context));
        return INSTANCE;
    }

    /**
     * Customized TrueTime Cache implementation.
     */
    public synchronized TrueTime withCustomizedCache(CacheInterface cacheInterface) {
        DISK_CACHE_CLIENT.enableCacheInterface(cacheInterface);
        return INSTANCE;
    }

    /**
     * clear the cached TrueTime info on device reboot.
     */
    public static void clearCachedInfo() {
        DISK_CACHE_CLIENT.clearCachedInfo();
    }

    public synchronized TrueTime withConnectionTimeout(int timeoutInMillis) {
        _udpSocketTimeoutInMillis = timeoutInMillis;
        return INSTANCE;
    }

    public synchronized TrueTime withRootDelayMax(float rootDelayMax) {
        if (rootDelayMax > _rootDelayMax) {
          String log = String.format(Locale.getDefault(),
              "The recommended max rootDelay value is %f. You are setting it at %f",
              _rootDelayMax, rootDelayMax);
          TrueLog.w(TAG, log);
        }

        _rootDelayMax = rootDelayMax;
        return INSTANCE;
    }

    public synchronized TrueTime withRootDispersionMax(float rootDispersionMax) {
      if (rootDispersionMax > _rootDispersionMax) {
        String log = String.format(Locale.getDefault(),
            "The recommended max rootDispersion value is %f. You are setting it at %f",
            _rootDispersionMax, rootDispersionMax);
        TrueLog.w(TAG, log);
      }

      _rootDispersionMax = rootDispersionMax;
      return INSTANCE;
    }

    public synchronized TrueTime withServerResponseDelayMax(int serverResponseDelayInMillis) {
        _serverResponseDelayMax = serverResponseDelayInMillis;
        return INSTANCE;
    }

    public synchronized TrueTime withNtpHost(String ntpHost) {
        _ntpHost = ntpHost;
        return INSTANCE;
    }

    public synchronized TrueTime withLoggingEnabled(boolean isLoggingEnabled) {
        TrueLog.setLoggingEnabled(isLoggingEnabled);
        return INSTANCE;
    }

    // -----------------------------------------------------------------------------------

    protected void initialize(String ntpHost) throws IOException {
        if (isInitialized()) {
            TrueLog.i(TAG, "---- TrueTime already initialized from previous boot/init");
            return;
        }

        requestTime(ntpHost);
        saveTrueTimeInfoToDisk();
    }

    long[] requestTime(String ntpHost) throws IOException {
        return SNTP_CLIENT.requestTime(ntpHost,
            _rootDelayMax,
            _rootDispersionMax,
            _serverResponseDelayMax,
            _udpSocketTimeoutInMillis);
    }

    synchronized static void saveTrueTimeInfoToDisk() {
        if (!SNTP_CLIENT.wasInitialized()) {
            TrueLog.i(TAG, "---- SNTP client not available. not caching TrueTime info in disk");
            return;
        }
        DISK_CACHE_CLIENT.cacheTrueTimeInfo(SNTP_CLIENT);
    }

    void cacheTrueTimeInfo(long[] response) {
        SNTP_CLIENT.cacheTrueTimeInfo(response);
    }

    private static long _getCachedDeviceUptime() {
        long cachedDeviceUptime = SNTP_CLIENT.wasInitialized()
                                  ? SNTP_CLIENT.getCachedDeviceUptime()
                                  : DISK_CACHE_CLIENT.getCachedDeviceUptime();

        if (cachedDeviceUptime == 0L) {
            throw new RuntimeException("expected device time from last boot to be cached. couldn't find it.");
        }

        return cachedDeviceUptime;
    }

    private static long _getCachedSntpTime() {
        long cachedSntpTime = SNTP_CLIENT.wasInitialized()
                              ? SNTP_CLIENT.getCachedSntpTime()
                              : DISK_CACHE_CLIENT.getCachedSntpTime();

        if (cachedSntpTime == 0L) {
            throw new RuntimeException("expected SNTP time from last boot to be cached. couldn't find it.");
        }

        return cachedSntpTime;
    }

}

File: library/src/main/java/com/instacart/library/truetime/TrueTime.kt
package com.instacart.library.truetime

import android.content.Context
import android.os.SystemClock
import java.io.IOException
import java.util.Date
import java.util.Locale

open class TrueTime {

    companion object {

        private val TAG = TrueTime::class.java.simpleName

        private val INSTANCE = TrueTime()
        private val DISK_CACHE_CLIENT = DiskCacheClient()
        private val SNTP_CLIENT = SntpClient()

        private var rootDelayMax = 100f
        private var rootDispersionMax = 100f
        private var serverResponseDelayMax = 750
        private var udpSocketTimeoutInMillis = 30000

        /**
         * @return Date object that returns the current time in the default Timezone
         */
        fun now(): Date {
            if (!isInitialized) {
                throw IllegalStateException("You need to call init() on TrueTime at least once.")
            }

            val cachedSntpTime = getCachedSntpTime()
            val cachedDeviceUptime = getCachedDeviceUptime()
            val deviceUptime = SystemClock.elapsedRealtime()
            val now = cachedSntpTime + (deviceUptime - cachedDeviceUptime)

            return Date(now)
        }

        val isInitialized: Boolean
            get() = SNTP_CLIENT.wasInitialized() || DISK_CACHE_CLIENT.isTrueTimeCachedFromAPreviousBoot()

        fun build(): TrueTime {
            return INSTANCE
        }

        /**
         * clear the cached TrueTime info on device reboot.
         */
        fun clearCachedInfo() {
            DISK_CACHE_CLIENT.clearCachedInfo()
        }

        @Synchronized fun saveTrueTimeInfoToDisk() {
            if (!SNTP_CLIENT.wasInitialized()) {
                TrueLog.i(TAG, "---- SNTP client not available. not caching TrueTime info in disk")
                return
            }
            DISK_CACHE_CLIENT.cacheTrueTimeInfo(SNTP_CLIENT)
        }

        private fun getCachedDeviceUptime(): Long {
            val cachedDeviceUptime = if (SNTP_CLIENT.wasInitialized())
                SNTP_CLIENT.cachedDeviceUptime
            else
                DISK_CACHE_CLIENT.getCachedDeviceUptime()

            if (cachedDeviceUptime == 0L) {
                throw RuntimeException("expected device time from last boot to be cached. couldn't find it.")
            }

            return cachedDeviceUptime
        }

        private fun getCachedSntpTime(): Long {
            val cachedSntpTime = if (SNTP_CLIENT.wasInitialized())
                SNTP_CLIENT.cachedSntpTime
            else
                DISK_CACHE_CLIENT.getCachedSntpTime()

            if (cachedSntpTime == 0L) {
                throw RuntimeException("expected SNTP time from last boot to be cached. couldn't find it.")
            }

            return cachedSntpTime
        }
    }

    private var _ntpHost = "1.us.pool.ntp.org"

    @Throws(IOException::class)
    fun initialize() {
        initialize(_ntpHost)
    }

    /**
     * Cache TrueTime initialization information in SharedPreferences
     * This can help avoid additional TrueTime initialization on app kills
     */
    @Synchronized open fun withSharedPreferencesCache(context: Context): TrueTime {
        DISK_CACHE_CLIENT.enableCacheInterface(SharedPreferenceCacheImpl(context))
        return INSTANCE
    }

    /**
     * Customized TrueTime Cache implementation.
     */
    @Synchronized open fun withCustomizedCache(cacheInterface: CacheInterface): TrueTime {
        DISK_CACHE_CLIENT.enableCacheInterface(cacheInterface)
        return INSTANCE
    }

    @Synchronized open fun withConnectionTimeout(timeoutInMillis: Int): TrueTime {
        udpSocketTimeoutInMillis = timeoutInMillis
        return INSTANCE
    }

    @Synchronized open fun withRootDelayMax(rootDelayMax: Float): TrueTime {
        if (rootDelayMax > rootDelayMax) {
            val log = String.format(
                Locale.getDefault(),
                "The recommended max rootDelay value is %f. You are setting it at %f",
                rootDelayMax, rootDelayMax
            )
            TrueLog.w(TAG, log)
        }

        TrueTime.rootDelayMax = rootDelayMax
        return INSTANCE
    }

    @Synchronized open fun withRootDispersionMax(rootDispersionMax: Float): TrueTime {
        if (rootDispersionMax > rootDispersionMax) {
            val log = String.format(
                Locale.getDefault(),
                "The recommended max rootDispersion value is %f. You are setting it at %f",
                rootDispersionMax, rootDispersionMax
            )
            TrueLog.w(TAG, log)
        }

        TrueTime.rootDispersionMax = rootDispersionMax
        return INSTANCE
    }

    @Synchronized open fun withServerResponseDelayMax(serverResponseDelayInMillis: Int): TrueTime {
        serverResponseDelayMax = serverResponseDelayInMillis
        return INSTANCE
    }

    @Synchronized fun withNtpHost(ntpHost: String): TrueTime {
        _ntpHost = ntpHost
        return INSTANCE
    }

    @Synchronized open fun withLoggingEnabled(isLoggingEnabled: Boolean): TrueTime {
        TrueLog.setLoggingEnabled(isLoggingEnabled)
        return INSTANCE
    }

    // -----------------------------------------------------------------------------------

    @Throws(IOException::class)
    protected fun initialize(ntpHost: String) {
        if (isInitialized) {
            TrueLog.i(TAG, "---- TrueTime already initialized from previous boot/init")
            return
        }

        requestTime(ntpHost)
        saveTrueTimeInfoToDisk()
    }

    @Throws(IOException::class)
    fun requestTime(ntpHost: String): LongArray {
        return SNTP_CLIENT.requestTime(
            ntpHost,
            rootDelayMax,
            rootDispersionMax,
            serverResponseDelayMax,
            udpSocketTimeoutInMillis
        )
    }

    fun cacheTrueTimeInfo(response: LongArray) {
        SNTP_CLIENT.cacheTrueTimeInfo(response)
    }
}
